### ISSUE TITLE
transitive trust based on blockchain constructs

### ISSUE THREAD

**Issue by synctext** [Tribler/tribler#2894]
```PhD Thesis direction```

By using the learn-by-doing methodology we will create a transitive trust solution.
Specifically we will design, implement, deploy, and incrementally improve transitive trust solution using our own append-only TrustChain solution.

Steps and dependancies:
- [x] True halves #2135 
- [x] live edges #2458 
- [ ] Gumby test coverage and deployment
- [ ] Remove Bloom filters and use Live Edge walking + record record + growing interval construct
![live_edge](https://cloud.githubusercontent.com/assets/325224/24652899/de5d25a0-1933-11e7-9bb3-f6f5640c4ee4.png)
- [x] Creat N-Hop-Sync across TrustChain to replace Full-Sync that fails to scale.
- [ ] Move TrustChain down from community level to replace discovery community. First, cut Dispersy into two : #2778
- [ ] Repair scalability of AllChannel with N-Hop-Sync

**Comment by pimveldhuisen** [comment]
Some work on live trust edges based in dispersy MultiChainCommunity, can be found on this branch:
https://github.com/pimveldhuisen/tribler/tree/trust-edges
Corresponding with this gumby branch:
https://github.com/pimveldhuisen/gumby/tree/trust-edges
And this DAS experiment:
https://jenkins.tribler.org/job/Tribler/job/Experiments/job/multichain/job/trust-walker_das/

**Comment by qstokkink** [comment]
Live edges implementation:
https://github.com/qstokkink/tribler/tree/live_edges
Live edges gumby experiment:
https://github.com/qstokkink/gumby/tree/live_edges

**Comment by qstokkink** [comment]
test

**Comment by qstokkink** [comment]
Isolated 1000 node DAS5 experiment is now up and running:

![trust_0hudoo2sq1dl564hzc41 7bdbsk](https://cloud.githubusercontent.com/assets/3630389/26276083/8e15b768-3d70-11e7-86fc-959eda4a717f.png)
![trust_0lex9dy13wfudbcf49y9f2hy42g](https://cloud.githubusercontent.com/assets/3630389/26276084/92b55120-3d70-11e7-8127-a1288f3b6fd7.png)
![trust_0lfvfjtsyhduckn7xft3v100g7a](https://cloud.githubusercontent.com/assets/3630389/26276085/98063fa4-3d70-11e7-9b5c-c93a53a73a7f.png)
![trust_02vdieqs9fm6gyb6fa skt a8by](https://cloud.githubusercontent.com/assets/3630389/26276086/9f0c9fb4-3d70-11e7-8688-605e1441356e.png)
![trust_074exgkpqtdtxgwj98skiikikkg](https://cloud.githubusercontent.com/assets/3630389/26276087/a406b6ee-3d70-11e7-9c1d-ad43c1ced438.png)

**Comment by qstokkink** [comment]
I spent yesterday creating an R package which allows me to easily make animated png files. This allows me to more easily visualize the evolution of a network in an automated manner. (Plus, animation is great for propaganda purposes)

https://github.com/qstokkink/apng

**Comment by qstokkink** [comment]
Fun update. My animated png R package is now part of the official R package repository: https://CRAN.R-project.org/package=apng

**Comment by synctext** [comment]
in R..

epic!

**Comment by devos50** [comment]
Nice! I've used your R script (with some modifications) to plot the transactions in my decentralized market experiment:

![transaction_graph](https://cloud.githubusercontent.com/assets/1707075/26522636/f22edef2-4304-11e7-8794-9100c3660b57.png)

**Comment by qstokkink** [comment]
The first (windowless/window size of 1) integration of UTP in Dispersy is now up and running: https://github.com/qstokkink/gumby/blob/utp_community/experiments/utp/utp_community.py

**Comment by synctext** [comment]
solid progress.
is there a protocol for 1UDP versus n-uTP transfers?

**Comment by qstokkink** [comment]
@synctext If I understood your question correctly, you want to know if multiple uTP transfers can be multiplexed over one UDP connection. If so: yes, you can have multiple uTP connections over a single UDP "link".

**Comment by synctext** [comment]
When two peers talk, do we have a protocol for indicating this is a single UDP Dispersy message or we now start setting up a uTP connection?
How to close? timeouts?

**Comment by qstokkink** [comment]
@synctext It is a single UDP Dispersy message, this makes it backwards compatible and allows it to make use of some of the Dispersy information shipped with it, without compromising the pure uTP protocol. The uTP connections are built on top of this Dispersy message. 

The extra information we can use, is the member we are exchanging information with. This  guarentees that there will be no _connection\_id_ collisions, so you will never accidentally (or on purpose) get someone else's data.

Closing the connections is done using the normal uTP _ST\_FIN_ message. In the worst case it will time-out (implemented with Twisted to avoid rogue threads and the like).

The interface for this functionality is as simple as:
```python
community.send_utp_message(candidate, data)
```
The implementation will then take care of setting up a connection and closing it once the data (a string) has been sent to the candidate.

Another possible use of this, is sending Protocol Buffers messages of arbitrary size (no longer restricted to a single UDP packet).

**Comment by synctext** [comment]
nice.
So what we want as a perfect API is protocol buffer wrapping with uTP transfer. I guess. But still we want statistics for a basic progress bar when a large binary blob is received.

**Comment by qstokkink** [comment]
@synctext Sure, we can hook in progress statistics for the sender and the receiver without any issues whatsoever.

Also: I hooked up windowing and retransmissions today. Once I document and clean the code a bit, Dispersy-uTP is ready for shipping. I'm looking forward to the thumbnail implementation 👍

**Comment by synctext** [comment]
nice
another key usage is magnet channels. We want to enable Tribler channels like:
https://www.youtube.com/user/CODblackopsPS
1+ million posted videos... At 100Bytes per video, a big channel blob... 100MB..

**Comment by qstokkink** [comment]
Somewhat unrelated: for fun, I implemented auto-googling for thumbnails of suggested torrents and channels today. 90% of the time the resulting thumbnail is actually pretty good.

![tribler_thumbs_torrents](https://cloud.githubusercontent.com/assets/3630389/26755136/aa888304-4887-11e7-8e70-e0814fabb740.png)

![tribler_thumbs_channels](https://cloud.githubusercontent.com/assets/3630389/26755142/b7311864-4887-11e7-8a4e-562b2c0a92a2.png)

<sub><sub>That one thumbnail will haunt me for the rest of my life.</sub></sub>

**Comment by synctext** [comment]
_related work_
[Real-Time Operationg system for Robotics with publish/subscribe](http://wiki.ros.org/Services), messages, and other great pieces of technology in distributed systems. Inspiration example of how good documentation is made. Makes us look like absolute amateurs :-)
Define the top-level [concepts](http://wiki.ros.org/ROS/Concepts) plus all related topics in one consistent place and a final [technical overview](http://wiki.ros.org/ROS/Technical%20Overview) with proper code snippets. 
Another [example of appealing looking docs](http://arbiterjs.com/#docs).

**Comment by qstokkink** [comment]
@synctext Functionally Dispersy can actually do more than that. The only thing we are missing in Dispersy is the fancy Arbiter system (which are pretty much Dispersy communities in disguise)* and TCP connections. However, as far as I can see, the user needs to set up his own TCP NAT rules in ROS; so no fancy TCP puncturing/hole punching in ROS either.

As far as docs go: touché, point taken. Their documentation is lightyears better than ours.

* Actually, we could even build such a system on top of the communities.

**Comment by qstokkink** [comment]
Alright, the uTP infrastructure has been refactored and documented (reminder: [here](https://github.com/qstokkink/gumby/blob/utp_community/experiments/utp/utp_community.py)) and available for deployment once we need it.

I will now move on to peer rediscovery/repuncturing. This will open the doors for all of the more fancy goals we want to achieve.

**Comment by qstokkink** [comment]
I'm starting the roll-out of live edges:

- [X] Dispersy PR - https://github.com/Tribler/dispersy/pull/548, status: READY
- [ ] Tribler PR - _code done, waiting for Dispersy PR_
- [ ] Gumby PR - _code done, waiting for Tribler PR_

**Comment by qstokkink** [comment]
While the live edges are being reviewed etc. I will move on to the n-hop-sync. The design is as follows:

## n-hop-sync

Extend functionality of `CommunityDestination`:

 - Instead of just having a fan-out for sharing this message (called [`node_count`](https://github.com/Tribler/dispersy/blob/devel/destination.py#L75) now), a `depth` parameter is added to determine (you guessed it) the sharing depth in the community.
 - [_Implementation detail_] When a `CommunityDestination` is decoded from the wire-format its depth is decremented, when it is forwarded (see [`Dispersy._forward()`](https://github.com/Tribler/dispersy/blob/devel/dispersy.py#L1249)) it is checked if the depth is greater than 0.

## n-trust-hop-sync

At this point I will probably be fed up with the fact that the destinations are [hardcoded](https://github.com/Tribler/dispersy/blob/devel/dispersy.py#L1275) and refactor the destination selection code into the `Destination` class, where it belongs. An extra `Destination` class will be created (called `TrustDestination`?) subclassing `CommunityDestination`, which will query the associated community through some `get_trust()` function to determine the candidates to send to.

**Comment by synctext** [comment]
Inspiring animations for live edges and graph sampling: https://bost.ocks.org/mike/algorithms/
![random_walk](https://user-images.githubusercontent.com/325224/28421867-b32fa07e-6d66-11e7-9386-1034ab76f4fb.png)

**Comment by synctext** [comment]
Related work on secure pairing, trust and discovery + remixing: https://github.com/solatis/bitgloom

**Comment by synctext** [comment]
More docs and related work on the Privacy+self-sovereign ticket: https://github.com/Tribler/tribler/issues/3013#issuecomment-313672996
<img src="https://user-images.githubusercontent.com/325224/30802366-915065e2-a1e6-11e7-80ec-3988c6985213.png" width="750">

**Comment by synctext** [comment]
[Learning trust strategies in reputation exchange networks](https://github.com/Tribler/tribler/files/1446510/learning_trust_strategies_in_reputation_exchange_networks.pdf)
```
This research proposes a method by which agents can learn trust
decision strategies using a q-learning technique. Q-learning [15]
is a well-known tool for learning best decision-making strategies
by associating actions with expected rewards. Several researchers
[3, 4, 9, 13, 19] have employed reinforcement learning techniques
to discover strategies for trust decisions. However, these trust
decisions only relate to fundamental transactions, as in the
Prisoner’s Dilemma [2]. Learning trust decision strategies in
reputation-transaction systems is significantly more difficult,
since in these systems, fundamental transaction decisions are not
independent from each other and from reputation transaction
decisions. An agent’s fundamental transactions with one partner
may influence its fundamental transactions with another, if its two
partners are able to communicate reputation information.
Learning an agent’s trust decision strategy is also complex
because multiple decisions can affect common results. For
example, falsely spreading negative reputations about a
trustworthy partner might cause other agents to trust the lying
agent while isolating the falsely accused agent.
```

**Comment by qstokkink** [comment]
Here's another (more up-to-date) list of Blockchain identity systems: https://github.com/peacekeeper/blockchain-identity

Also people are apparently starting to catch up to our ideas:
https://link.springer.com/chapter/10.1007/978-3-319-67816-0_21

**Comment by synctext** [comment]
We might have an alternative algorithm to find the non-sybil region of Trustchain by using an NP-hard graph search algorithm. The maximum k-cluster in a DAG might indicate the largest honest region. [phantom algorithm was presented by researcher Yonatan Sompolinsky](http://www.cs.huji.ac.il/~yoni_sompo/) at BPASE'18 event. His earlier work on DAGchain:
"SPECTRE: Serialization of Proof-of-work Events: Confirming Transactions via Recursive Elections".

**Comment by synctext** [comment]
Great Princeton work: [SybilFuse: Combining Local Attributes with Global Structure to Perform Robust Sybil Detection](http://www.princeton.edu/~pmittal/publications/sybilfuse-mis2)

**Comment by synctext** [comment]
University College London: ["Decentralizing Public Key Infrastructures with ClaimChains"](https://arxiv.org/pdf/1707.06279.pdf). Very similar design idea to our deployed technology; has Merkle tree and fancy skip list for efficiency.
```
A ClaimChain consists of blocks, each block containing
all the information necessary to represent the claims of
one user about her own keys, and her beliefs about other
users’ ClaimChains (which we call cross references).
Users can access each other’s ClaimChains and combine
these beliefs to support the provision of evidence about
the binding between identities and keys. We note that
the design of ClaimChains could handle any generic
claims about user beliefs. However, for the sake of simplicity,
in this document we restrict ourselves to the
context of key management for messaging applications.
A core goal of our design is to support full decentralization.
Yet, ClaimChains are flexible in terms of
deployment.
```

**Comment by qstokkink** [comment]
This issue has been superseded by other issues. Live edge walking has been in Tribler for over a year now and there are other issues for trustworthy gossip.

**Comment by qstokkink** [comment]
Post-mortem: trust-based walking was a very bad idea.

**Comment by synctext** [comment]
Post-mortem2: load-balancing needs to be maintained when doing non-random walks! Trust introduces a strong bias. Destructive and leads to overloading of highly-trusted nodes in a power-law setting or complete neglect of low-trust nodes.
